rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,400 @@
+
+There already exists components with similar purposes in Bevy:
+[`Interaction` and `FocusPolicy`](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ui/src/focus.rs#L14).
+However, it only supports a barebone mouse-oriented navigation scheme. And the
+focus handling is not even implemented.
+
+My proposal, on top of making it easy to implement controller navigation of
+in-game UIs, enables keyboard navigation and answers accessibility concerns
+such as the ones
+[raised in this issue](https://github.com/bevyengine/bevy/issues/254).
+
+## User-facing explanation
+
+Examples and an implementation-oriented description is available in the
+[ui-navigation plugin README](https://github.com/nicopap/ui-navigation/blob/master/Readme.md).
+
+To interact in real-time with the UI, the developer emits `NavRequest`s events.
+The navigation system responds with `NavEvent` events, such as
+`NavEvent::FocusChanged` or `NavEvent::Caught`. 
+
+The navigation system is the only one to handle the focus state of the UI. The
+user interacts with it strictly by sending `NavRequest`s, reading `NavEvent` or
+using the `Focusable` methods. Typically in a system with a
+`Query<&Focusable, Changed<Focusable>>` parameter.
+
+Rather than directly reacting to input, we opt to make our system reactive to
+`NavRequest`, this way the developer is free to chose which button does what in
+the game. The developer also has the complete freedom to swap at runtime what
+sends or how those `NavRequest` are generated.
+
+### Example
+
+Let's start with a practical example, and try figure out how we could specify
+navigation in it:
+
+![Typical RPG tabbed menu](https://user-images.githubusercontent.com/26321040/140716885-eb089626-21d2-4711-a0c9-cf185bc0f19a.png)
+
+The tab "soul" and the panel "abc" are highlighted to show the player what menu
+they are down, they are **active** elements. The circle "B" is highlighted
+differently to show it is the **focused** element.
+
+You can move `UP`/`DOWN` in the submenu, but you can also use `LT`/`RT` to
+go from one submenu to the other (the _soul_, _body_ etc tabs). In code, we
+might react to focus changes (for example to change the color of
+focused/unfocused elements) as follow:
+```rust
+fn setup_ui() {
+  // setup UI like you would setup UI in bevy today
+}
+fn handle_nav_events(mut events: EventReader<NavEvent>, game: Res<Gameui>) {
+    use bevy_ui_navigation::{NavEvent::Caught, NavRequest::Action};
+    for event in events.iter() {
+        match event {
+            Caught { from, request: Action } if from.first() == game.start_game_button => {
+                  // Start the game on "A" or "ENTER" button press
+            }
+            _ => {}
+        }
+    }
+}
+fn button_system(
+    materials: Res<Materials>,
+    mut focusables: Query<(&Focusable, &mut Handle<ColorMaterial>), Changed<Focusable>>,
+) {
+    for (focus_state, mut material) in focusables.iter_mut() {
+        if focus_state.is_focused() {
```

Comment:
Does this make more sense as a `match` on an enum? I tend to find that to be clearer, and it helps me catch edge cases more easily.

Reply:
It's implemented as an `enum` under the hood! Makes total sense to expose it then.